aws-athena/views/reporting-vw_assessment_roll_muni.sql

```diff
@@ -0,0 +1,114 @@
+empty City of Chicago/class 4 row to exist for the mailed stage. */
+WITH stages AS (
+
+    SELECT 'mailed' AS stage
+    UNION
+    SELECT 'assessor certified' AS stage
+    UNION
+    SELECT 'bor certified' AS stage
+
+),
+
+/* This CTE removes historical PINs in reporting.vw_pin_township_class that are
+not in reporting.vw_pin_value_long. We do this to make sure the samples we
+derive the numerator and denominator from for stage_portion are the same. These
+differences are inherent in the tables that feed these views (iasworld.pardat
+and iasworld.asmt_all, respectively) and are data errors - not emblematic of
+what portion of a municipality has actually progressed through an assessment
+stage.
+
+It does NOT remove PINs from the most recent year of
+reporting.vw_pin_township_class since we expect these differences based on how
+iasworld.asmt_all is populated through the year as the assessment cycle
+progresses.
+
+Starting in 2020 a small number of PINs are present in iasworld.asmt_all for
+one or two but not all three stages of assessment when we would expect all three
+stages to be present for said PINs. This is also a data error, but is NOT
+addressed in this view and leads to a few instances where stage_portion ends up
+being less than 1 when it should equal 1. 16-07-219-029-1032 missing a mailed
+value but having CCAO and BOR certified values in 2021 is an example. */
+trimmed_town_class AS (
+    SELECT vptc.*
+    FROM {{ ref('reporting.vw_pin_township_class') }} AS vptc
+    LEFT JOIN
+        (
+            SELECT DISTINCT
+                pin,
+                year
+            FROM {{ ref('reporting.vw_pin_value_long') }}
+        )
+            AS pins
+        ON vptc.pin = pins.pin
+        AND vptc.year = pins.year
+    WHERE pins.pin IS NOT NULL
+        OR vptc.year
+        = (SELECT MAX(year) FROM {{ ref('reporting.vw_pin_township_class') }})
+
+),
+
+/* Calculate the denominator for the stage_portion column.
+reporting.vw_pin_township_class serves as the universe of yearly PINs we expect
+to see in reporting.vw_pin_value_long. */
+pin_counts AS (
+    SELECT
+        vptc.municipality_name,
+        vptc.major_class,
+        vptc.year,
+        stages.stage,
+        COUNT(*) AS total_n
+    FROM trimmed_town_class AS vptc
+    CROSS JOIN stages
+    WHERE vptc.municipality_name IS NOT NULL
+    GROUP BY
+        vptc.municipality_name,
+        vptc.year,
+        vptc.major_class,
+        stages.stage
+)
+
+-- Calculate total and median values by municipality, as well as the portion of
+-- each municipality that has progressed through an assessment stage by class.
+SELECT
+    pin_counts.year,
+    pin_counts.stage,
+    pin_counts.municipality_name,
+    munis.major_class AS class,
+    SUM(CAST(vpvl.pin IS NOT NULL AS INT)) AS n,
+    pin_counts.total_n,
+    SUM(CAST(vpvl.pin IS NOT NULL AS DOUBLE))
+    / CAST(pin_counts.total_n AS DOUBLE) AS stage_portion,
```

Comment:
suggestion: I find these column names slightly ambiguous. Maybe instead:

- n --> num_pin_w_value
- total_n --> num_pin_total_in_group
- stage_portion --> pct_pin_w_value_in_group

Reply:
I'm cool with this.